### PR TITLE
Steps to Care Future Needs

### PR DESCRIPTION
**Pull request by nateh777**
##### What does the change add or fix?

Add the capability to schedule future needs and display them. In order to do so all the assignment methods were moved to a common utility class instead of being located in the Care Entry Block.

**New Settings:**

Care Dashboard
- **Threshold of Days For Future Need**, The number of days the need has to be within before displaying on the dashboard.

Care Entry   
- **Threshold of Days before Assignment**, The number of days you can schedule a need in the future before a need will be assigned to workers.

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

- Added capability to schedule future needs in Steps to Care.

---------

##### Who reported, requested, or paid for the change?

RSC

---------

##### Does this update or add options to the block UI?

Care Dashboard, "Include Future Needs"
image

image

Care Entry, new settings:
image

---------

##### What files does it affect?

- StepsToCare/CareDashboard.ascx
- StepsToCare/CareDashboard.ascx.cs
- StepsToCare/CareEntry.ascx.cs

---------

##### Is it a breaking change for other versions/clients?

Yes, new DLL update is required with new utility methods. https://github.com/KingdomFirst/RockAssemblies/pull/83

### CODE REVIEW

File: StepsToCare/CareDashboard.ascx.cs [hunk 1]
```diff
@@ -210,6 +210,12 @@ namespace RockWeb.Plugins.rocks_kfs.StepsToCare
         Order = 19,
         Key = AttributeKey.NoteViewLavaTemplate )]
 
+    [IntegerField( "Threshold of Days For Future Need",
+        Description = "The number of days the need has to be within before displaying on the dashboard.",
```
Comment (nateh777) [inline StepsToCare/CareDashboard.ascx.cs:214]:
```suggestion
    [IntegerField( "Threshold of Days for Scheduled Needs",
        Description = "The number of days from today to display scheduled needs on the dashboard.",
```

File: StepsToCare/CareDashboard.ascx [hunk 2]
```diff
@@ -109,6 +111,7 @@
                             <Rock:RockDropDownList ID="ddlSubmitter" runat="server" Label="Submitted By" EnhanceForLongLists="true" />
                             <Rock:CampusPicker ID="cpCampus" runat="server" Label="Campus" />
                             <Rock:RockCheckBox ID="cbAssignedToMe" runat="server" Label="Assigned to Me" />
+                            <Rock:RockCheckBox ID="cbIncludeFutureNeeds" runat="server" Label="Include Future Needs" />
```
Comment (nateh777) [inline StepsToCare/CareDashboard.ascx:114]:
```suggestion
                            <Rock:RockCheckBox ID="cbIncludeFutureNeeds" runat="server" Label="Include Scheduled Needs" />
```

File: StepsToCare/CareDashboard.ascx.cs [hunk 3]
```diff
@@ -270,6 +277,7 @@ private static class UserPreferenceKey
             public const string Status = "Status";
             public const string Campus = "Campus";
             public const string AssignedToMe = "Assigned to Me";
+            public const string IncludeFutureNeeds = "Include Future Needs";
```
Comment (nateh777) [inline StepsToCare/CareDashboard.ascx.cs:280]:
```suggestion
            public const string IncludeScheduledNeeds = "Include all Scheduled Needs";
```

File: StepsToCare/CareDashboard.ascx.cs [hunk 4]
```diff
@@ -571,6 +579,7 @@ protected void rFilter_ApplyFilterClick( object sender, EventArgs e )
             rFilter.SaveUserPreference( UserPreferenceKey.Status, "Status", dvpStatus.SelectedItem.Value );
             rFilter.SaveUserPreference( UserPreferenceKey.Campus, "Campus", cpCampus.SelectedCampusId.ToString() );
             rFilter.SaveUserPreference( UserPreferenceKey.AssignedToMe, "Assigned to Me", cbAssignedToMe.Checked.ToString() );
+            rFilter.SaveUserPreference( UserPreferenceKey.IncludeFutureNeeds, "Include Future Needs", cbIncludeFutureNeeds.Checked.ToString() );
```
Comment (nateh777) [inline StepsToCare/CareDashboard.ascx.cs:582]:
```suggestion
            rFilter.SaveUserPreference( UserPreferenceKey.IncludeScheduledNeeds, "Include all Scheduled Needs", cbIncludeFutureNeeds.Checked.ToString() );
```

File: StepsToCare/CareDashboard.ascx.cs [hunk 5]
```diff
@@ -1966,6 +1979,11 @@ private void BindMainGrid( RockContext rockContext, CareNeedService careNeedServ
                 qry = qry.Where( b => b.DateEntered <= endDate );
             }
 
+            if ( !cbIncludeFutureNeeds.Checked )
+            {
+                qry = qry.Where( b => b.DateEntered <= futureDate );
```
Comment (nateh777) [inline StepsToCare/CareDashboard.ascx.cs:1984]: Update date range picker to restrict end date to today.

File: StepsToCare/CareEntry.ascx.cs [hunk 6]
```diff
@@ -97,14 +93,18 @@ namespace RockWeb.Plugins.rocks_kfs.StepsToCare
         Key = AttributeKey.AdultFamilyWorkers,
         Category = AttributeCategory.FamilyNeeds )]
 
+    [IntegerField( "Threshold of Days before Assignment",
+        Description = "The number of days you can schedule a need in the future before a need will be assigned to workers.",
```
Comment (nateh777) [inline StepsToCare/CareEntry.ascx.cs:97]:
```suggestion
        Description = "The number of days before a scheduled need is assigned to workers. Default: 3",
```

File: StepsToCare/CareDashboard.ascx.cs (hunk 5 above)
Comment (nateh777) [reply StepsToCare/CareDashboard.ascx.cs:1984]: Added end date to datepicker in this commit: https://github.com/KingdomFirst/RockBlocks/pull/114/commits/965f2ccf4d6bd32001336fd09f7ca7acbd51bdb1

**Review by mattbaylor — APPROVED** [review]

**Pull request merged**